lib/phoenix_live_view/upload_channel.ex

```diff
@@ -0,0 +1,131 @@
+defmodule Phoenix.LiveView.UploadChannel do
+  @moduledoc false
+  use Phoenix.Channel, log_handle_in: false
+
+  require Logger
+
+  alias Phoenix.LiveView.{Static, Channel}
+
+  def cancel(pid) do
+    GenServer.call(pid, :cancel)
+  end
+
+  def consume(pid, entry, func) when is_function(func, 1) or is_function(func, 2) do
+    GenServer.call(pid, {:consume, entry, func})
+  end
+
+  @impl true
+  def join(_topic, auth_payload, socket) do
+    %{"token" => token} = auth_payload
+
+    with {:ok, %{pid: pid, ref: ref}} <- Static.verify_token(socket.endpoint, token),
+         {:ok, config} <- Channel.register_upload(pid, ref),
+         %{max_file_size: max_file_size, chunk_timeout: chunk_timeout} = config,
+         {:ok, path} <- Plug.Upload.random_file("live_view_upload"),
+         {:ok, handle} <- File.open(path, [:binary, :write]) do
+      Process.monitor(pid)
+
+      socket =
+        assign(socket, %{
+          path: path,
+          handle: handle,
+          live_view_pid: pid,
+          max_file_size: max_file_size,
+          chunk_timeout: chunk_timeout,
+          chunk_timer: nil,
+          done?: false,
+          uploaded_size: 0
+        })
+
+      {:ok, socket}
+    else
+      {:error, :limit_exceeded} -> {:error, %{reason: :limit_exceeded}}
+      _ -> {:error, %{reason: "invalid_token"}}
+    end
+  end
+
+  @impl true
+  def handle_in("chunk", {:binary, payload}, socket) do
+    %{uploaded_size: uploaded_size, max_file_size: max_file_size} = socket.assigns
+    socket = reschedule_chunk_timer(socket)
+
+    if byte_size(payload) + uploaded_size <= max_file_size do
+      {:reply, :ok, write_bytes(socket, payload)}
+    else
+      reply = %{reason: "file_size_limit_exceeded", limit: max_file_size}
+      {:stop, {:shutdown, :closed}, {:error, reply}, socket}
+    end
+  end
+
+  @impl true
+  def handle_info(
+        {:DOWN, _, _, live_view_pid, reason},
+        %{assigns: %{live_view_pid: live_view_pid}} = socket
+      ) do
+    reason = if reason == :normal, do: {:shutdown, :closed}, else: reason
+    {:stop, reason, socket}
+  end
+
+  def handle_info(:chunk_timeout, socket) do
+    {:stop, {:shutdown, :closed}, socket}
+  end
+
+  @impl true
+  def handle_call({:consume, entry, func}, from, socket) do
+    unless socket.assigns.done?,
+      do: raise(RuntimeError, "cannot consume uploaded file that is still in progress")
```

Comment:
Instead of raising here, which would load to a confusing error message, what if we reply with `{:error, :in_progress}` and raise this on the caller?